packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -6278,50 +6281,197 @@ abstract class RenderObjectElement extends Element {
+              'KeepAliveParentDataMixin.'
+            ),
+            ErrorDescription(
+              'The ownership chain for the RenderObject that received the '
+              'parent data was:\n  ${debugGetCreatorChain(10)}'
+            ),
           ]);
-        } on FlutterError catch (e) {
-          _reportException(ErrorSummary('while looking for parent data.'), e, e.stackTrace);
+        } on FlutterError catch (error) {
+          _reportException(
+            ErrorSummary('while looking for parent data.'),
+            error,
+            error.stackTrace,
+          );
         }
       }
       return true;
     }());
+  }
+
+  List<ParentDataElement<ParentData>> _findAncestorParentDataElements() {
+    Element? ancestor = _parent;
+    final List<ParentDataElement<ParentData>> result = <ParentDataElement<ParentData>>[];
+    final Set<Type> debugAncestorTypes = <Type>{};
+    final Set<Type> debugParentDataTypes = <Type>{};
+    final List<Type> debugAncestorCulprits = <Type>[];
+    final List<Type> debugParentDataCulprits = <Type>[];
+
+    // More than one ParentDataWidget can contribute ParentData, but there are
+    // some constraints.
+    // 1. ParentData can only be written by unique ParentDataWidget types.
+    //    For example, two KeepAlive ParentDataWidgets trying to write to the
+    //    same child is not allowed.
+    // 2. Each contributing ParentDataWidget must contribute to a unique
+    //    ParentData type, less ParentData be overwritten.
+    //    For example, there cannot be two ParentDataWidgets that both write
+    //    ParentData of type KeepAliveParentDataMixin, if the first check was
+    //    subverted by a subclassing of the KeepAlive ParentDataWidget.
+    // 3. The ParentData itself must be compatible with all ParentDataWidgets
+    //    writing to it.
+    //    For example, TwoDimensionalViewportParentData uses the
+    //    KeepAliveParentDataMixin, so it could be compatible with both
+    //    KeepAlive, and another ParentDataWidget with ParentData type
+    //    TwoDimensionalViewportParentData or a subclass thereof.
+    // The first and second cases are verified here. The third is verified in
+    // debugIsValidRenderObject.
+
+    while (ancestor != null && ancestor is! RenderObjectElement) {
+      if (ancestor is ParentDataElement<ParentData>) {
+        assert(() {
+          if (!debugAncestorTypes.add(ancestor.runtimeType)) {
+            if (!debugAncestorCulprits.contains(ancestor.runtimeType)) {
+              // Add the first one we had put in the Set of Types.
+              debugAncestorCulprits.add(ancestor.runtimeType);
+            }
+            debugAncestorCulprits.add(ancestor.runtimeType);
+          }
+
+          if (!debugParentDataTypes.add((ancestor! as ParentDataElement<ParentData>).parentDataType)) {
+            if (!debugParentDataCulprits.contains((ancestor as ParentDataElement<ParentData>).parentDataType)) {
```

Comment:
Is Dart insisting on both of these explicit casts? I would have assumed that the `if` in line 6434 promotes ancestor? Or maybe that promotion doesn't survive in the assert closure?

Reply:
It is! I was going to file a bug about it. I'll link it back here once I do.

Reply:
Follow up, filed https://github.com/dart-lang/sdk/issues/53602